Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/security/EncryptionErrorResetTool.kt

```diff
@@ -0,0 +1,137 @@
+        set(value) = prefs.edit {
+            putBoolean(PKEY_EA1851_SHOW_RESET_NOTICE, value)
+        }
+
+    fun tryResetIfNecessary(error: Throwable): Boolean {
+        Timber.w(error, "isRecoveryWarranted()")
+
+        // We only reset for the first error encountered after upgrading to 1.4.0+
+        if (isResetWindowConsumed) {
+            Timber.v("Reset window has been consumed -> no reset.")
+            return false
+        }
+        isResetWindowConsumed = true
+
+        val keyException = error.causes().singleOrNull { it is GeneralSecurityException }
+        if (keyException == null) {
+            Timber.v("Error has no GeneralSecurityException as cause -> no reset.")
+            return false
+        }
+
+        // https://github.com/google/tink/blob/a8ec74d083068cd5e1ebed86fd8254630617b592/java_src/src/main/java/com/google/crypto/tink/aead/AeadWrapper.java#L83
+        if (keyException.message?.trim()?.equals("decryption failed") != true) {
+            Timber.v("Not the GeneralSecurityException we are looking for -> no reset.")
+            return false
+        }
+
+        if (!encryptedPreferencesFile.exists()) {
+            // The error we are looking for can only happen if there already is an encrypted preferences file
+            Timber.w(
+                "Error fits, but where is the existing preference file (%s)? -> no reset.",
+                encryptedPreferencesFile
+            )
+            return false
+        }
+
+        // So we have a GeneralSecurityException("decryption failed") and existing preferences
+        // And this is the first error we encountered after upgrading to 1.4.0, let's do this!
+
+        return try {
+            performReset()
+        } catch (e: Exception) {
+            // If anything goes wrong, we return false, so that our caller can rethrow the original error.
+            Timber.e(e, "Error while performing the reset.")
+            false
+        }
+    }
+
+    private fun performReset(): Boolean {
+        // Delete encrypted shared preferences file
+
+        if (!encryptedPreferencesFile.delete()) {
+            Timber.w("Couldn't delete %s", encryptedPreferencesFile)
+            // The encrypted preferences are a prerequisite for this error case
+            // If we can't delete that, we have to assume our reset approach failed.
+            return false
+        }
+
+        // The user may have encountered the error even before a database was created.
+        if (encryptedDatabaseFile.exists() && !encryptedDatabaseFile.delete()) {
+            Timber.w("Couldn't delete %s", encryptedDatabaseFile)
+            // There was a database, but we couldn't delete it
+            // The database is inaccessible without the matching preferences (which we just deleted).
+            return false
+        }
+
+        resetPerformedAt = Instant.now()
+        isResetNoticeToBeShown = true
+
+        return true
+    }
+
+    companion object {
```

Comment:
to top of class

Reply:
We attempt to follow the official Kotlin conventions, which say `companion object` last: https://kotlinlang.org/docs/reference/coding-conventions.html